usr.sbin/makefs/zfs.c

```diff
@@ -74,7 +74,13 @@ zfs_prep_opts(fsinfo_t *fsopts)
 	zfs_opt_t *zfs;
 	size_t align;
 
-	align = alignof(uint64_t);
+	/*
+	 * The first field of zfs_opt_t must be aligned.
+	 *
+	 * For some reason snmalloc returns EINVAL if the alignment is smaller
```

Comment:
This does feel like a bug in snmalloc.  If the minimum alignment it can handle is sizeof(void *) it should just silently round up requested alignments to that.  It still fulfills the caller's intent to return 16 byte alignment for an 8 byte alignment request.

Reply:
Humm, posix_memalign requires a power of 2 greater than or equal to 'sizeof(void *)', so this isn't snmalloc specific, just POSIX.

Reply:
C11 says:

> The value of alignment shall be a valid alignment supported by the implementation and the value of size shall be an integral multiple of alignment.

where:

> Alignments are represented as values of the type size_t. Valid alignments include only those values returned by an _Alignof expression for fundamental types, plus an additional implementation-defined set of values, which may be empty. Every valid alignment value shall be a nonnegative integral power of two.

It sounds like snmalloc's aligned_alloc is inheriting the following constraint from posix_memalign:

> The value of alignment shall be a power of two multiple of sizeof(void *).

Reply:
Ah, so really I should fix this upstream then. I'll do that and bring this back as a cherry-pick.

Reply:
Filed https://github.com/microsoft/snmalloc/issues/668

Reply:
So to be clear, it seems like aligned_alloc() should allow sizes smaller than void *, but posix_memalign() does not.

jemalloc sets the min_alignment to 1 for aligned_alloc, and sizeof(void *) for posix_memalign(), so it does seem to be a snmalloc bug.

Reply:
Yes

Reply:
The snmalloc fix has been merged upstream into snmalloc.  I think this PR can probably be closed, though we will want to update snmalloc at some point?

Reply:
Yeah

Reply:
Thank you! I'll just carry the makefs patch as a local diff for now.